gitlab/resource_gitlab_project_freeze_period.go

```diff
@@ -0,0 +1,148 @@
+package gitlab
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+func resourceGitlabProjectFreezePeriod() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGitlabProjectFreezePeriodCreate,
+		Read:   resourceGitlabProjectFreezePeriodRead,
+		Update: resourceGitlabProjectFreezePeriodUpdate,
+		Delete: resourceGitlabProjectFreezePeriodDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"project_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"freeze_start": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"freeze_end": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"cron_timezone": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "UTC",
+			},
+		},
+	}
+}
+
+func resourceGitlabProjectFreezePeriodCreate(d *schema.ResourceData, meta interface{}) error {
+	projectID := d.Get("project_id").(string)
+
+	options := gitlab.CreateFreezePeriodOptions{
+		FreezeStart:  gitlab.String(d.Get("freeze_start").(string)),
+		FreezeEnd:    gitlab.String(d.Get("freeze_end").(string)),
+		CronTimezone: gitlab.String(d.Get("cron_timezone").(string)),
+	}
+
+	log.Printf("[DEBUG] Project %s create gitlab project-level freeze period %+v", projectID, options)
+
+	client := meta.(*gitlab.Client)
+	FreezePeriod, _, err := client.FreezePeriods.CreateFreezePeriodOptions(projectID, &options)
+	if err != nil {
+		return err
+	}
+
+	FreezePeriodIDString := fmt.Sprintf("%d", FreezePeriod.ID)
+	d.SetId(buildTwoPartID(&projectID, &FreezePeriodIDString))
+
+	return resourceGitlabProjectFreezePeriodRead(d, meta)
+}
+
+func resourceGitlabProjectFreezePeriodRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	projectID, freezePeriodID, err := projectIDAndFreezePeriodIDFromID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] read gitlab FreezePeriod %s/%d", projectID, freezePeriodID)
+
+	freezePeriod, _, err := client.FreezePeriods.GetFreezePeriod(projectID, freezePeriodID)
```

Comment:
You should handle a 404 error here and call `d.SetId("")`

Example: https://github.com/gitlabhq/terraform-provider-gitlab/blob/5dfb4cc39fc17489868406d0be770894b8d89d0d/gitlab/resource_gitlab_instance_cluster.go#L156-L164

Reply:
@armsnyder I don't see that code in master? Nor do I see any function called `is404`?

Reply:
Correct, it's not in master yet. `is404` was added in that PR too, so don't worry about using it.

Here's one way of checking the status code, from an error object:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/ef0ddea75d64794283f181dce06c6395f6dfdfe6/gitlab/data_source_gitlab_project.go#L195-L196

Reply:
Here's an example from master which includes the `setId("")` part.

https://github.com/gitlabhq/terraform-provider-gitlab/blob/ef0ddea75d64794283f181dce06c6395f6dfdfe6/gitlab/resource_gitlab_group_membership.go#L86-L94

Reply:
@armsnyder Gotcha, I've added the check.